DB/src/main/java/io/deephaven/db/v2/StreamTableToAppendOnlyTable.java

```diff
@@ -0,0 +1,82 @@
+     */
+    public static Table streamToAppendOnlyTable(Table streamTable) {
+        return QueryPerformanceRecorder.withNugget("streamToAppendOnlyTable", () -> {
+            Object streamAttribute = streamTable.getAttribute(Table.STREAM_TABLE_ATTRIBUTE);
+            if (!(streamAttribute instanceof Boolean) || !(Boolean) streamAttribute) {
+                throw new IllegalArgumentException("Input is not a stream table!");
+            }
+
+            final Map<String, ArrayBackedColumnSource> columns = new LinkedHashMap<>();
+            final int columnCount = streamTable.getColumnSourceMap().size();
+            final ColumnSource [] sourceColumns = new ColumnSource[columnCount];
+            final ArrayBackedColumnSource [] destColumns = new ArrayBackedColumnSource[columnCount];
+            int colIdx = 0;
+            for (Map.Entry<String, ? extends ColumnSource> nameColumnSourceEntry : streamTable.getColumnSourceMap().entrySet()) {
+                final ColumnSource<?> existingColumn = nameColumnSourceEntry.getValue();
+                final ArrayBackedColumnSource<?> newColumn = ArrayBackedColumnSource.getMemoryColumnSource(0, existingColumn.getType(), existingColumn.getComponentType());
+                columns.put(nameColumnSourceEntry.getKey(), newColumn);
+                sourceColumns[colIdx] = existingColumn;
+                destColumns[colIdx++] = newColumn;
+            }
+
+            Index index = Index.FACTORY.getEmptyIndex();
+            final QueryTable result = new QueryTable(index, columns);
+            result.setRefreshing(true);
+            result.setAttribute(Table.ADD_ONLY_TABLE_ATTRIBUTE, true);
+            result.setFlat();
+
+            ((DynamicTable) streamTable).listenForUpdates(new BaseTable.ShiftAwareListenerImpl("streamToAppendOnly", (DynamicTable) streamTable, result) {
+                @Override
+                public void onUpdate(Update upstream) {
+                    if (upstream.modified.nonempty() || upstream.shifted.nonempty()) {
+                        throw new IllegalArgumentException("Stream tables should not modify or shift!");
+                    }
+                    final long newRows = upstream.added.size();
+                    if (newRows == 0) {
+                        return;
+                    }
+                    final long currentSize = index.size();
+                    columns.values().forEach(c -> c.ensureCapacity(currentSize + newRows));
+
+                    final Index newRange = Index.FACTORY.getIndexByRange(currentSize, currentSize + newRows - 1);
+
+                    ChunkUtils.copyData(sourceColumns, upstream.added, destColumns, newRange, false);
+                    index.insertRange(currentSize, currentSize + newRows - 1);
+
+                    final ShiftAwareListener.Update downstream = new ShiftAwareListener.Update();
+                    downstream.shifted = IndexShiftData.EMPTY;
```

Comment:
What kind of crazy person thinks of these fields in this order?

Reply:
The kind that puts the empty crap that is needed for initialization first, and then the one thing you actually care about.